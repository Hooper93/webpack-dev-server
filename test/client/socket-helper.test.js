'use strict';

describe('socket', () => {
  afterEach(() => {
    jest.resetAllMocks();
    jest.resetModules();
  });

  it('should default to WebsocketClient when no clientClass provided', () => {
    jest.mock('../../client/clients/WebsocketClient');
    const socket = require('../../client/default/socket');
    const WebsocketClient = require('../../client/clients/WebsocketClient');

    const mockHandler = jest.fn();
    socket(
      'my.url',
      {
        example: mockHandler,
      },
      null
    );

    const mockClientInstance = WebsocketClient.mock.instances[0];

    // this simulates receiving a message from the server and passing it
    // along to the callback of onMessage
    mockClientInstance.onMessage.mock.calls[0][0](
      JSON.stringify({
        type: 'example',
        data: 'hello world',
      })
    );

    expect(WebsocketClient.mock.calls[0]).toMatchSnapshot();
    expect(mockClientInstance.onOpen.mock.calls).toMatchSnapshot();
    expect(mockClientInstance.onClose.mock.calls).toMatchSnapshot();
    expect(mockClientInstance.onMessage.mock.calls).toMatchSnapshot();
    expect(mockHandler.mock.calls).toMatchSnapshot();
  });

  it('should use clientClass when passed in', () => {
    // sockjs is simply being used here so that the class can be
    // mocked and we can tell a WebsocketClient instance is not created
    jest.mock('../../client/clients/WebsocketClient');
    jest.mock('../../client/clients/SockJSClient');
    const socket = require('../../client/default/socket');
    const WebsocketClient = require('../../client/clients/WebsocketClient');
    const SockJSClient = require('../../client/clients/SockJSClient');
    const mockClientClass = SockJSClient;

    const mockHandler = jest.fn();
    socket(
      'my.url',
      {
        example: mockHandler,
      },
      mockClientClass
    );

    const mockClientInstance = mockClientClass.mock.instances[0];

    // this simulates receiving a message from the server and passing it
    // along to the callback of onMessage
    mockClientInstance.onMessage.mock.calls[0][0](
      JSON.stringify({
        type: 'example',
        data: 'hello world',
      })
    );

    expect(WebsocketClient.mock.instances.length).toEqual(0);
    expect(mockClientClass.mock.calls[0]).toMatchSnapshot();
    expect(mockClientInstance.onOpen.mock.calls).toMatchSnapshot();
    expect(mockClientInstance.onClose.mock.calls).toMatchSnapshot();
    expect(mockClientInstance.onMessage.mock.calls).toMatchSnapshot();
    expect(mockHandler.mock.calls).toMatchSnapshot();
  });
});
