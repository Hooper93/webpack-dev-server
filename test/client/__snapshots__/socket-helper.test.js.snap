// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`socket should default to WebsocketClient when no clientClass provided 1`] = `
Array [
  "my.url",
]
`;

exports[`socket should default to WebsocketClient when no clientClass provided 2`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`socket should default to WebsocketClient when no clientClass provided 3`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`socket should default to WebsocketClient when no clientClass provided 4`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`socket should default to WebsocketClient when no clientClass provided 5`] = `
Array [
  Array [
    "hello world",
  ],
]
`;

exports[`socket should use clientClass when passed in 1`] = `
Array [
  "my.url",
]
`;

exports[`socket should use clientClass when passed in 2`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`socket should use clientClass when passed in 3`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`socket should use clientClass when passed in 4`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`socket should use clientClass when passed in 5`] = `
Array [
  Array [
    "hello world",
  ],
]
`;
